### ISSUE TITLE
CaptureScreenshotWindow("...", ImGuiCaptureFlags_StitchAll ) fails

### ISSUE THREAD

**Issue by pthom** [ocornut/imgui_test_engine#33]
Re-Bonjour Omar,

Sorry if I'm flooding the issues a bit today. 

I'm trying the code of app_minimal, and I have an issue with screen capture and the flag ImGuiCaptureFlags_StitchAll.

Basically, 

* I'm getting a fully black capture under windows (although it has the correct size). This is strange. I'm using Windows 10 ARM 64 bits (although I compiled the code in x64 mode)
*  under macOS (after having applied a hack to take into account the FrameBufferScale), the scroll does not work,  and I get a repeated image:

<img width="231" alt="image" src="https://github.com/ocornut/imgui_test_engine/assets/7694091/533636f8-6e89-49cf-8bef-91ad5df345cc">

i.e. the size is correct, but it seems like the content is repeated (as if the scroll did not work). 
Based on the `usleep(1000)` for linux, I tried to add a sleep(1 second) during the capture, and I could see that effectively the scroll does not work.

**Comment by pthom** [comment]
> I have however now confirmed that it breaks with io.ConfigWindowsMoveFromTitleBarOnly = true in the same way as your repeated capture. Seems fixable.

Good catch! I if set io.ConfigWindowsMoveFromTitleBarOnly = false, then it works
<img width="170" alt="image" src="https://github.com/ocornut/imgui_test_engine/assets/7694091/93c996eb-7b03-40ec-88f1-d8fc00fec9d7">

I was getting this issue from inside HelloImGui, where io.ConfigWindowsMoveFromTitleBarOnly is false by default. I'll soon post an issue inside HelloImGui and give some details about how it is going. If you want to give me some output it would be appreciable! 

Now, getting on to debug under windows

**Comment by pthom** [comment]
Under windows, I have found the issue: I'm using the docking branch, and this code inside app_minimal_main.cpp is the reason it fails:

```cpp
#ifdef IMGUI_HAS_VIEWPORT
    io.ConfigFlags |= ImGuiConfigFlags_ViewportsEnable;
#endif
```

If I comment out this line, then it works! I think your implementation mentions somewhere that viewports are not handled for screen capture.

**Comment by ocornut** [comment]
- I fixed the issue with `io.ConfigWindowsMoveFromTitleBarOnly` and stitching.

- ImGuiCaptureFlags_StitchAll doesn't currently work with multi-viewport because the way we move the window makes it move to its own viewport whereas we try to capture from the main viewport. I however pushed changes to support capturing from secondary viewport in imgui_app (the API was already there but not supported by imgui_app), but it requires further changes in capture tool too.

**Comment by pthom** [comment]
> I fixed the issue with io.ConfigWindowsMoveFromTitleBarOnly and stitching

I confirm that your fix works

**Comment by ocornut** [comment]
I put a little more effort on trying to fix various things with the capture tool, but I believe the codebase is too flawed, it may be simpler to rewrite with specialized/separate function for single capture, stitched capture, video capture etc. and test support for viewport and DPI shenanigans.

**Comment by pthom** [comment]
It happens to all of us :-) 

May be adding simple structs to store a capture and its coords such as below could help to make the API a bit simpler.

```cpp
struct ImageBufferRgba
{
    int Width = 0, Height = 0;
    ImVector<uint8_t> Buffer;
};

struct CaptureCoords
{
   ImGuiID ViewportId;
   int x, int y, int w, int h; // Or maybe ImRect
}
```


with the added benefits that:
* transforming CaptureCoords into new coords taking into account FrameBufferScale now becomes easier
* resizing ImageBufferRgba.Buffer can be done when needed (i.e. at the moment of the capture). Signalling that the image buffer size differs from the window size becomes also possible.

And the callback signature could then become:
```cpp
typedef bool (ImGuiScreenCaptureFunc)(CaptureCoords coords, ImageBufferRgba* buffer, void* user_data);
```

However, this would probably change too much user code for those who already implemented a callback.

**Comment by pthom** [comment]
... and there are probably lots of other reasons for which this would not work  ;-)

**Comment by pthom** [comment]
FYI, I started to work on integrating imgui_test_engine into hello_imgui (which is a preliminary, before integrating into Dear ImGui Bundle).

If you are interested, I made a quick demo on how easy it can be to setup a test app here:
 https://github.com/pthom/hello_imgui_test_engine_demo

At the moment, the integration of imgui_test_engine is not in the main branch of HelloImGui, but in the with_imgui_test_engine branch. 

I'll make sure to add some doc about its license when I release it in the main branch.

The code for integratings the test engine was quite simple to add, and is grouped in [this folder](https://github.com/pthom/hello_imgui/tree/with_imgui_test_engine/src/hello_imgui_test_engine_integration).
